Remove `ecs` option on `user_agent` processor

PeteGillinElastic opened the pull request:
This removes the `ecs` option on the `user_agent` ingest processor, which is deprecated (way back in 6.7) and ignored. It will no longer be possible to create instances with this option, and the option will be removed from instances persisted in the cluster state on startup.

The mechanism to do this upgrade on startup is designed to be reusable for other upgrades either to ingest processors or more generally to any custom metadata. It is inspired by the existing mechanism to upgrade index templates.

PeteGillinElastic committed:
Remove `ecs` option on `user_agent` processor

This removes the `ecs` option on the `user_agent` ingest processor,
which is deprecated and ignored. It was deprecated way back in 6.7.0,
so no BWC mode is needed.

elasticsearchmachine commented:
Hi @PeteGillinElastic, I've created a changelog YAML for you. *Note* that since this PR is labelled `>breaking`, you need to update the changelog YAML to fill out the extended information sections.

PeteGillinElastic committed:
Update docs/changelog/116077.yaml

PeteGillinElastic committed:
Update changelog

PeteGillinElastic committed:
Fix breaking.area in changelog

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into v9-user-agent-process-ecs

PeteGillinElastic committed:
Add code to upgrade cluster state on start-up by removing the `ecs`
parameter on `user-agent` ingest processors.

PeteGillinElastic committed:
Add implementation comment, revert accidentaly API change

PeteGillinElastic committed:
Clarify changelog

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into v9-user-agent-process-ecs

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

PeteGillinElastic commented on modules/ingest-user-agent/src/main/java/org/elasticsearch/ingest/useragent/UserAgentProcessor.java:
```diff
@@ -272,4 +260,11 @@ public static Property parseProperty(String propertyName) {
         }
     }
+
+    static boolean maybeUpgradeConfig(Map<String, Object> config) {
```
Question: Are we expected to support upgrades which skip a major version? If not, I think we can marked this for removal in v10. This is because it's not possible to add processors with the deprecated parameter in v9, and any added in v8 or earlier will have been fixed by the upgrade to v9.

thecoop replied:
All upgrades to vMAJOR have to go through the last minor of the previous major (v[MAJOR-1].LAST)

elasticsearchmachine commented:
Pinging @elastic/es-core-infra (Team:Core/Infra)

thecoop commented on server/src/main/java/org/elasticsearch/ingest/IngestMetadata.java:
```diff
@@ -169,4 +169,39 @@ public boolean equals(Object o) {
+    /**
+     * Functional interface for upgrading processor configs. An implementation of this will be associated with a specific processor type.
+     */
+    public interface ProcessorConfigUpgrader {
```
Could this just be `Predicate<Map<String, Object>>`?

thecoop replied:
I suppose it begs the question 'why is this _not_ a `Predicate`?'. The mutation is a good reason to have a separate type, though

thecoop commented on server/src/main/java/org/elasticsearch/ingest/PipelineConfiguration.java:
```diff
@@ -156,4 +156,40 @@ public int hashCode() {
+        if (processorList instanceof Iterable) {
+            for (Object processorMap : (Iterable<?>) processorList) {
+                if (processorMap instanceof Map) {
+                    Object processorConfig = ((Map<?, ?>) processorMap).get(type);
```
You can use `XContentMapValues.extractValue` here

thecoop replied:
Maybe its just all the casting then. You can use `if (processorList instanceof Iterable<?> it) { for (Object m : it) {...}` to make it a bit more readable.

ldematte commented on server/src/test/java/org/elasticsearch/gateway/GatewayMetaStateTests.java:
```diff
@@ -98,11 +100,37 @@ public void testCustomMetadataNoChange() {
+        CustomMetadata1 originalCustom1 = new CustomMetadata1("old data");
+        CustomMetadata1 upgradedCustom1 = new CustomMetadata1("new data");
+        CustomMetadata2 custom2 = new CustomMetadata2("some data");
+        Metadata originalMetadata = randomMetadata(originalCustom1, custom2);
```
I have troubles understading this test: it seems we are testing randomly either one custom metadata type or the other?

PeteGillinElastic replied:
Yeah, I can see how this is confusing, sorry! In fact, `randomMetadata` isn't introducing any randomness to do with the customs, only in other irrelevant stuff. I'll replace it with something more explicitly showing what needs to be created for the test, and add a comment or two.

ldematte commented:
Overall looks good, thanks! But I've a doubt on how this works with multiple custom metadata - maybe I've just not understood correctly how it works/what's possible to combine and what not. Tests did not help me understand neither, unfortunately.

PeteGillinElastic committed:
Respond to review comments

PeteGillinElastic reviewed (commented):
Thanks for the review! I've responded to everything and pushed some changes, PTAL.

PeteGillinElastic committed:
spotlessApply

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into v9-user-agent-process-ecs

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into v9-user-agent-process-ecs

ldematte commented on server/src/main/java/org/elasticsearch/gateway/GatewayMetaState.java:
```diff
@@ -300,18 +300,31 @@ static Metadata upgradeMetadata(Metadata metadata, IndexMetadataVerifier indexMe
+            if (original != null) {
+                Metadata.Custom upgraded = upgrader.apply(original);
+                if (upgraded.equals(original) == false) {
+                    upgradedMetadata.putCustom(type, upgraded);
```
I tried to look at the test and got confused (see my other comment). 
My concern was that we started with an empty metadata and upgradedMetadata onlu contained the new, upgraded (not equal) Metadata.Custom. But if that is initialized with a copy of the original metadata, it's OK.

thecoop commented on server/src/main/java/org/elasticsearch/plugins/MetadataUpgrader.java:
```diff
@@ -15,21 +15,30 @@
         };
+        this.customMetadataUpgraders = customMetadataUpgraders.stream()
+            .flatMap(map -> map.entrySet().stream())
+            .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue, Function::andThen));
```
Any reason this uses `andThen` (which will cause a chain of functions), rather than using the iterative approach used for index template upgraders? Both should use the same style - either functions, or iterative.

PeteGillinElastic replied:
Yeah, fair point. I have changed it to something which iterates over the operators to combine them. It's still not the exactly the same as the templates one, because this one is dealing with a map of operators rather than an operator over maps — this is by design, as here each operator only wants to touch a single custom metadata type. Hopefully this makes sense! I'm including some comments 'cos the collector stuff is a little hairy. I'm also improving the tests in the same commit.

PeteGillinElastic committed:
Tidy up PipelineConfiguration.maybeUpgradeProcessors using pattern-matching instanceof, as suggested in review

thecoop commented on server/src/main/java/org/elasticsearch/ingest/PipelineConfiguration.java:
```diff
@@ -156,4 +156,40 @@ public int hashCode() {
+            try (XContentBuilder builder = XContentBuilder.builder(xContentType.xContent())) {
+                return new PipelineConfiguration(id, BytesReference.bytes(builder.map(mutableConfigMap)), xContentType);
+            } catch (IOException e) {
+                throw new RuntimeException(e);
```
Can this be `UncheckedIOException` instead?

PeteGillinElastic replied:
Done.

thecoop reviewed (approved):
A few code style changes to make, but core structure LGTM

ldematte reviewed (approved):
Tests looks more clear to me now, thanks!
Changes to add the reusable mechanism to do upgrades (to Plugins and MetadataUpgrader) looks good to me. I'll leave Ingest and UserAgent related changes to DataManagement.

PeteGillinElastic committed:
Change `GatewayMetaState` to combine custom upgraders by iterating over them rather then composing them, as suggested in review

Also strengthen the unit test (and refactor it to be more like a fake
than a mock while we're at it).

PeteGillinElastic committed:
Merge remote-tracking branch 'upstream/main' into v9-user-agent-process-ecs

PeteGillinElastic committed:
More specific wrapper exception type

PeteGillinElastic reviewed (commented):
Thanks again both! I've done everything as I best understand it.

PeteGillinElastic committed:
Make a concrete class to combine operators

dakrone reviewed (approved):
LGTM, thanks for adding this mechanism Pete. It would be great if we had a 7.x state we could stick somewhere and "recover" from in the test, but I don't think it's a blocker for this work to go in.

PeteGillinElastic commented:
I think this depends on the policy for the serverless changelog. This property was deprecated and ignored since 6.x, so nobody should have been using it in serverless. I don't know whether this should be mentioned or not.

Also, this change was made in November, before serverless GA.
